File: Proyecto_P1.Lib/Repositories/VehiculoRepository.cs
using Proyecto_P1.Lib.Models;
using System;
using System.Collections.Generic;

namespace Proyecto_P1.Lib.Repositories
{
    /// <summary>
    /// Implementación del repositorio de vehículos usando arreglos primitivos.
    /// Redimensionamiento manual y desplazamiento de elementos.
    /// </summary>
    public class VehiculoRepository : IVehiculoRepository
    {
        private Vehiculo[] _vehiculos;
        private int _cantidad;

        public VehiculoRepository()
        {
            _vehiculos = new Vehiculo[10]; // Tamaño inicial
            _cantidad = 0;
        }

        public int Cantidad => _cantidad;

        public void Insertar(Vehiculo vehiculo)
        {
            if (vehiculo == null)
                throw new ArgumentNullException(nameof(vehiculo), "El vehículo no puede ser nulo.");

            // Redimensionar si el arreglo está lleno
            if (_cantidad == _vehiculos.Length)
            {
                Vehiculo[] nuevoArreglo = new Vehiculo[_vehiculos.Length * 2];
                for (int i = 0; i < _cantidad; i++)
                {
                    nuevoArreglo[i] = _vehiculos[i];
                }
                _vehiculos = nuevoArreglo;
            }

            _vehiculos[_cantidad] = vehiculo;
            _cantidad++;
        }

        public void EliminarPorIndice(int indice)
        {
            if (indice < 0 || indice >= _cantidad)
                throw new IndexOutOfRangeException($"Índice {indice} fuera de rango. Rango válido: 0 a {_cantidad - 1}.");

            // Desplazar elementos a la izquierda para cubrir el hueco
            for (int i = indice; i < _cantidad - 1; i++)
            {
                _vehiculos[i] = _vehiculos[i + 1];
            }
            
            _vehiculos[_cantidad - 1] = null!; // Limpiar última referencia
            _cantidad--;
        }

        public void EliminarPorPlaca(string placa)
        {
            int indiceEncontrado = -1;
            for (int i = 0; i < _cantidad; i++)
            {
                if (_vehiculos[i].Placa.Equals(placa, StringComparison.OrdinalIgnoreCase))
                {
                    indiceEncontrado = i;
                    break;
                }
            }

            if (indiceEncontrado == -1)
                throw new KeyNotFoundException($"No se encontró un vehículo con la placa '{placa}'.");

            EliminarPorIndice(indiceEncontrado);
        }

        public void Modificar(int indice, Vehiculo vehiculoActualizado)
        {
            if (indice < 0 || indice >= _cantidad)
                throw new IndexOutOfRangeException($"Índice {indice} fuera de rango. Rango válido: 0 a {_cantidad - 1}.");

            if (vehiculoActualizado == null)
                throw new ArgumentNullException(nameof(vehiculoActualizado), "El vehículo actualizado no puede ser nulo.");

            _vehiculos[indice] = vehiculoActualizado;
        }

        public Vehiculo[] ObtenerTodos()
        {
            Vehiculo[] resultado = new Vehiculo[_cantidad];
            for (int i = 0; i < _cantidad; i++)
            {
                resultado[i] = _vehiculos[i];
            }
            return resultado;
        }

        public Vehiculo? BuscarPorPlaca(string placa)
        {
            for (int i = 0; i < _cantidad; i++)
            {
                if (_vehiculos[i].Placa.Equals(placa, StringComparison.OrdinalIgnoreCase))
                {
                    return _vehiculos[i];
                }
            }
            return null;
        }

        public Vehiculo? BuscarPorIndice(int indice)
        {
            if (indice < 0 || indice >= _cantidad)
                return null;

            return _vehiculos[indice];
        }

        public void LimpiarTodo()
        {
            _vehiculos = new Vehiculo[10];
            _cantidad = 0;
        }
    }
}

File: ProyectoEstructuras1/src/main/java/com/mycompany/proyectoestructuras1/VehiculoRepository.java
package com.mycompany.proyectoestructuras1;

/**
 * Implementación del repositorio de vehículos usando arreglos primitivos.
 * Redimensionamiento manual y desplazamiento de elementos.
 * Patrón Singleton para compartir datos entre formularios.
 */
public class VehiculoRepository implements IVehiculoRepository {

    private static VehiculoRepository instancia;

    private Vehiculo[] vehiculos;
    private int cantidad;

    private VehiculoRepository() {
        vehiculos = new Vehiculo[10]; // Tamaño inicial
        cantidad = 0;
    }

    public static VehiculoRepository getInstancia() {
        if (instancia == null) {
            instancia = new VehiculoRepository();
        }
        return instancia;
    }

    @Override
    public int getCantidad() {
        return cantidad;
    }

    @Override
    public void insertar(Vehiculo vehiculo) {
        if (vehiculo == null) {
            throw new IllegalArgumentException("El vehículo no puede ser nulo.");
        }

        // Redimensionar si el arreglo está lleno
        if (cantidad == vehiculos.length) {
            Vehiculo[] nuevoArreglo = new Vehiculo[vehiculos.length * 2];
            for (int i = 0; i < cantidad; i++) {
                nuevoArreglo[i] = vehiculos[i];
            }
            vehiculos = nuevoArreglo;
        }

        vehiculos[cantidad] = vehiculo;
        cantidad++;
    }

    @Override
    public void eliminarPorIndice(int indice) {
        if (indice < 0 || indice >= cantidad) {
            throw new IndexOutOfBoundsException(
                    "Índice " + indice + " fuera de rango. Rango válido: 0 a " + (cantidad - 1) + ".");
        }

        // Desplazar elementos a la izquierda para cubrir el hueco
        for (int i = indice; i < cantidad - 1; i++) {
            vehiculos[i] = vehiculos[i + 1];
        }

        vehiculos[cantidad - 1] = null; // Limpiar última referencia
        cantidad--;
    }

    @Override
    public void eliminarPorPlaca(String placa) {
        int indiceEncontrado = -1;
        for (int i = 0; i < cantidad; i++) {
            if (vehiculos[i].getPlaca().equalsIgnoreCase(placa)) {
                indiceEncontrado = i;
                break;
            }
        }

        if (indiceEncontrado == -1) {
            throw new IllegalArgumentException("No se encontró un vehículo con la placa '" + placa + "'.");
        }

        eliminarPorIndice(indiceEncontrado);
    }

    @Override
    public void modificar(int indice, Vehiculo vehiculoActualizado) {
        if (indice < 0 || indice >= cantidad) {
            throw new IndexOutOfBoundsException(
                    "Índice " + indice + " fuera de rango. Rango válido: 0 a " + (cantidad - 1) + ".");
        }

        if (vehiculoActualizado == null) {
            throw new IllegalArgumentException("El vehículo actualizado no puede ser nulo.");
        }

        vehiculos[indice] = vehiculoActualizado;
    }

    @Override
    public Vehiculo[] obtenerTodos() {
        Vehiculo[] resultado = new Vehiculo[cantidad];
        for (int i = 0; i < cantidad; i++) {
            resultado[i] = vehiculos[i];
        }
        return resultado;
    }

    @Override
    public Vehiculo buscarPorPlaca(String placa) {
        for (int i = 0; i < cantidad; i++) {
            if (vehiculos[i].getPlaca().equalsIgnoreCase(placa)) {
                return vehiculos[i];
            }
        }
        return null;
    }

    @Override
    public Vehiculo buscarPorIndice(int indice) {
        if (indice < 0 || indice >= cantidad) {
            return null;
        }
        return vehiculos[indice];
    }

    @Override
    public void limpiarTodo() {
        vehiculos = new Vehiculo[10];
        cantidad = 0;
    }
}
